tests/python/unittest/test_gen_requirements.py

```diff
@@ -0,0 +1,220 @@
+]
+
+
+def test_validate_requirements():
+    with patch(gen_requirements, REQUIREMENTS_BY_PIECE=None):
+        assert gen_requirements.validate_requirements_by_piece() == [
+            "must be list or tuple, see None"
+        ]
+
+    with patch(gen_requirements, REQUIREMENTS_BY_PIECE=PROBLEM_REQUIREMENTS):
+        problems = gen_requirements.validate_requirements_by_piece()
+        assert problems == [
+            'piece extras-pre-core: must list after "core" (core must be first)',
+            "piece extras-pre-core: deps should be a list of strings, got ['foo', 123]",
+            "piece 456: must be str",
+            "piece wrong-description-type: description should be a string, got None",
+            (
+                'piece bad-value: should be formatted like ("bad-value", ("<requirements.txt '
+                'comment>", ["dep1", "dep2", ...])). got: None'
+            ),
+            (
+                'piece bad-value-2: should be formatted like ("bad-value-2", '
+                '("<requirements.txt comment>", ["dep1", "dep2", ...])). got: (\'\', '
+                "['foo'], 34)"
+            ),
+            "piece invalid: listed twice",
+            "piece invalid: deps should be a list of strings, got ['baz', None, 123]",
+            "piece unsorted: deps must be sorted. Correct order:\n  ['bar', 'foo', 'qux']",
+            "piece versioned_dep: deps must be sorted. Correct order:\n  ['bar>4', 'baz==1.2', 'buz<3', 'foo==^2.0']",
+            "piece versioned_dep: dependency baz==1.2 should not specify a version. Add it to CONSTRAINTS instead.",
+            "piece versioned_dep: dependency foo==^2.0 should not specify a version. Add it to CONSTRAINTS instead.",
+            "piece versioned_dep: dependency buz<3 should not specify a version. Add it to CONSTRAINTS instead.",
+            "piece versioned_dep: dependency bar>4 should not specify a version. Add it to CONSTRAINTS instead.",
+            "piece duplicate_dep: dependency buz listed twice",
+            'piece extras-post-dev: must list before "dev" (dev must be last)',
+            'pieces other than "core" and "dev" must appear in alphabetical order: '
+            "['bad-value', 'bad-value-2', 'duplicate_dep', 'extras-foo', 'extras-post-dev', "
+            "'extras-pre-core', 'invalid', 'invalid', 'unsorted', 'versioned_dep', "
+            "'wrong-description-type']",
+        ]
+
+
+TEST_REQUIREMENTS_BY_PIECE = (
+    ("core", ("core tvm requirements", ("bar", "foo", "non-constrained"))),
+    ("extra-one", ("requirements for one feature", ("baz", "qux"))),
+    ("extra-two", ("requirements for two feature", ("buz", "qux", "semver-minor", "semver-patch"))),
+    ("dev", ("requirements for dev", ("buz", "oof", "rab"))),
+)
+
+
+def test_validate_constraints():
+    with patch(
+        gen_requirements,
+        REQUIREMENTS_BY_PIECE=TEST_REQUIREMENTS_BY_PIECE,
+        CONSTRAINTS=(
+            ("unlisted", "~=3"),
```

Comment:
Same comment as above, I think it would be better if the problems were kept along with the input constraints that are being tested.

Reply:
same issue though